### PR TITLE
Userspace: muppetjones (#1)

### PR DESCRIPTION
**Pull request by muppetjones**
## Description

This PR adds:

- My userspace
- Planck keymap
- Lily58 keymap
- Kyria keymap

My userspace and keymaps incorporate code from the following users:

- @splitkb (optimizations)
- @drashna (`wrapper.h` pattern)
- @andrewjrae (casemodes)
- @j-inc (bongo cat)

I've added the following features in my userspace and keymaps. As I clean up and optimize these features, I'll add them into the main code base.

- Etch-a-Mouse: Encoder-based pointing device movement with acceleration.
- Dynamic, Layer-based RGB: RGB underglow that changes with your layers and respects your current RGB settings. (Need to move this from my kyria into my userspace).

This PR only includes one change outside of my user-specific code: Updates to `keyboards/lily58/lib/layer_state_reader.c`. The default Lily58 keymap did not compile without these updates.

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [x] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Issues Fixed or Closed by This PR

* 

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: keyboards/lily58/lib/layer_state_reader.c
Comment (zvecr) [inline keyboards/lily58/lib/layer_state_reader.c:7]: These changes would break the current assumptions for the default keymaps and any user ones based off it.

File: users/muppetjones/rules.mk
Comment (zvecr) [inline users/muppetjones/rules.mk:6]: You want to be using the "official" way, here and elsewhere, instead
```suggestion
LTO_ENABLE = yes
```

**Review by zvecr — CHANGES_REQUESTED** [review]

File: keyboards/kyria/keymaps/muppetjones/rules.mk
Comment (drashna) [inline keyboards/kyria/keymaps/muppetjones/rules.mk:14]:
```suggestion
```

`LTO_ENABLE = yes` does all of this, actually.

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:12]: This isn't needed, it is automatically added. 
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:98]: This will return the value as a string: 
```suggestion
    uint8_t n = get_current_wpm();
    char wpm_counter[4];
    wpm_counter[3] = '\0';
    wpm_counter[2] = '0' + n % 10;
    wpm_counter[1] = ( n /= 10) % 10 ? '0' + (n) % 10 : (n / 10) % 10 ? '0' : ' ';
    wpm_counter[0] = n / 10 ? '0' + n / 10 : ' ';
```

And will use ~1.5kb less

File: keyboards/lily58/keymaps/muppetjones/keymap.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/keymap.c:27]:
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/keymap.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/keymap.c:37]:
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/rules.mk
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/rules.mk:11]:
```suggestion
LTO_ENABLE = yes
```

File: keyboards/lily58/keymaps/muppetjones/config.h
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/config.h:56]:
```suggestion
```

File: users/muppetjones/config.h
Comment (drashna) [inline users/muppetjones/config.h:1]: Could you add a GPL2+ compatible license header here? 

For instance: 
https://github.com/qmk/qmk_firmware/blob/28ff51175b6d45ce5a8dbfe7ed6e7a12df3bc8a8/quantum/template/base/keyboard.c#L1-L15

File: users/muppetjones/features/etchamouse.h
Comment (drashna) [inline users/muppetjones/features/etchamouse.h:20]:
```suggestion
#if defined(POINTING_DEVICE_ENABLE) && defined(ENCODER_ENABLE)
```

File: users/muppetjones/features/etchamouse.h
Comment (drashna) [inline users/muppetjones/features/etchamouse.h:61]:
```suggestion
#endif
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (muppetjones) [reply keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:98]: Nice! Thanks!

File: keyboards/lily58/lib/layer_state_reader.c
Comment (muppetjones) [inline keyboards/lily58/lib/layer_state_reader.c:35]: Due to `clang-format`.

File: keyboards/lily58/lib/layer_state_reader.c
Comment (muppetjones) [reply keyboards/lily58/lib/layer_state_reader.c]: Rolled these back.

**Comment by muppetjones** [comment]
@drashna @zvecr Made suggested changes. Thanks for the review!

File: keyboards/lily58/keymaps/muppetjones/config.h
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/config.h:38]: Not used. 
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:11]:
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:33]:
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c:103]:
```suggestion
    return wpm_counter;
```

**Review by drashna — CHANGES_REQUESTED** [review]

**Review by drashna — APPROVED** [review]

File: keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c
Comment (muppetjones) [reply keyboards/lily58/keymaps/muppetjones/features/bongo_cat.c]: I actually just updated this last night while updating my lily58 keymap!

**Comment by muppetjones** [comment]
@drashna Thank you for the second review. All of the comments have been addressed.

I updated and changed some things while trying to debug a [ghosting problem](https://www.reddit.com/r/olkb/comments/ofmzgz/need_help_with_ghosting_problem/?). I saw the behaviour on my kyria initially, but also when I flashed my lily with the original updates I posted here. The keys ghosting were different on the lily; however, it was still multiple keys at once. The issue seems to be in QMK as the WPM display was always high -- hanging around 128-255.

I was originally using 0.11.* with no issue, but I had pulled in the latest master (0.13.*) to get my fork up to speed. Rolling back the QMK version and the chibios version to the most recent 0.11 fixed both the ghosting and the WPM.

All of my changes actually line up with the second round of requested changes. I haven't retested on 0.13 yet. I'm going to get my testing and dev setup a little more stable and see if I can't track down the source of the above issue. If I find anything, I'll open a new PR.

**Comment by drashna** [comment]
Yeah, there is an issue with the WPM code right now.  IIRC, there is a proposed fix for that. 

Also, develop seems to handle the split code better.

File: keyboards/lily58/keymaps/muppetjones/keymap.c
Comment (fauxpark) [inline keyboards/lily58/keymaps/muppetjones/keymap.c:24]: Fairly sure this isn't needed.
```suggestion
```

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (fauxpark) [inline keyboards/planck/keymaps/muppetjones/config.h:17]:
```suggestion
```

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (fauxpark) [inline keyboards/planck/keymaps/muppetjones/config.h:34]:
```suggestion
```

File: keyboards/lily58/keymaps/muppetjones/rules.mk
Comment (fauxpark) [inline keyboards/lily58/keymaps/muppetjones/rules.mk:25]:
```suggestion
WPM_ENABLE = yes
TAP_DANCE_ENABLE = yes
LTO_ENABLE = yes
```

File: keyboards/lily58/keymaps/muppetjones/rules.mk
Comment (fauxpark) [inline keyboards/lily58/keymaps/muppetjones/rules.mk:41]:
```suggestion
```

File: keyboards/planck/keymaps/muppetjones/keymap.c
Comment (fauxpark) [inline keyboards/planck/keymaps/muppetjones/keymap.c:258]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/muppetjones/keymap.c
Comment (muppetjones) [reply keyboards/lily58/keymaps/muppetjones/keymap.c:24]: If I recall, it didn't compile without it.

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (muppetjones) [reply keyboards/planck/keymaps/muppetjones/config.h:17]: Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (muppetjones) [reply keyboards/planck/keymaps/muppetjones/config.h:34]: Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

File: keyboards/lily58/keymaps/muppetjones/rules.mk
Comment (muppetjones) [reply keyboards/lily58/keymaps/muppetjones/rules.mk:25]: Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

This is also from the default lily58 keymap.

File: keyboards/planck/keymaps/muppetjones/keymap.c
Comment (muppetjones) [reply keyboards/planck/keymaps/muppetjones/keymap.c:258]: Leaving this was intentional, and as comment lines aren't included at compile time, it doesn't hurt to leave them in a personal keymap.

**Comment by muppetjones** [comment]
Not going to lie, I'm not a fan of requested changes that have zero effect on the code base (and one that actually affects compile negatively). All but one (the colors in the planck) are sections from the default keymap for that board that I had left for one reason or another. Had these been changes submitted that hit the actual codebase, I certainly would have cleaned those up beforehand, and could understand the requested changes if I had missed them.

I'm all for constructive criticism, but removing inconsequential comments and "requiring changes"...it just slows down and frustrates the review process. I've been on the reviewer side in that situation several times. A word of advice: Just "comment" instead of "request changes". If it really is critical to have the extra comments removed, please say so.

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (fauxpark) [reply keyboards/planck/keymaps/muppetjones/config.h:17]: This define no longer exists.

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (fauxpark) [reply keyboards/planck/keymaps/muppetjones/config.h:34]: This define no longer exists.

File: keyboards/lily58/keymaps/muppetjones/rules.mk
Comment (fauxpark) [reply keyboards/lily58/keymaps/muppetjones/rules.mk:25]: The idea here is to reduce duplication. Since most of this stuff is the same as the keyboard-level rules.mk, it can be removed, leaving the keymap-level rules.mk containing only the changes vs. keyboard-level.

File: keyboards/planck/keymaps/muppetjones/keymap.c
Comment (fauxpark) [reply keyboards/planck/keymaps/muppetjones/keymap.c:258]: Why duplicate what the docs already provide?

**Comment by fauxpark** [comment]
@muppetjones it's preferable to get things like this done *before* the PR is merged and is potentially copy-pasted into other users' code, bloating the repo. It's then harder for us to clean it up afterwards, because we generally require signoff from the affected users (or at least provide a heads up) when modifying their code.

**Comment by muppetjones** [comment]
@fauxpark That's good reasoning. I appreciate the response. I'll clean it up, and I'll see about another PR to clean up the defaults that I copied from.

File: keyboards/planck/keymaps/muppetjones/keymap.c
Comment (muppetjones) [reply keyboards/planck/keymaps/muppetjones/keymap.c:258]: I copied from there. I was experimenting, and wanted a more direct reference so I didn't have to look it up each time. I figured that since it was a comment, it wasn't hurting anything.

This experiment culminated in a dynamic layer lighting that I haven't moved into my user space yet.

**Comment by muppetjones** [comment]
@fauxpark @zvecr I've made the requested changes -- thanks for the patience and the feedback.

**Comment by muppetjones** [comment]
Bump for re-review when you get a chance. Thanks!
@fauxpark @zvecr

File: keyboards/lily58/keymaps/muppetjones/rules.mk
Comment (fauxpark) [inline keyboards/lily58/keymaps/muppetjones/rules.mk:4]:
```suggestion
```

**Review by fauxpark — CHANGES_REQUESTED** [review]
You have submodule changes.

**Comment by muppetjones** [comment]
@fauxpark Fixed.

**Comment by muppetjones** [comment]
Bump @fauxpark @zvecr

**Review by drashna — CHANGES_REQUESTED** [review]

**Comment by muppetjones** [comment]
@drashna Thanks for the re-review and changes. I'd kinda given up on this PR =]

I'll take a look and try to get this updated and cleaned up this weekend.

File: keyboards/planck/keymaps/muppetjones/config.h
Comment (muppetjones) [reply keyboards/planck/keymaps/muppetjones/config.h:1]:
```suggestion
 /* Copyright 2020 Stephen Bush <muppetjones>
  * 
  * This program is free software: you can redistribute it and/or modify 
  * it under the terms of the GNU General Public License as published by 
  * the Free Software Foundation, either version 2 of the License, or 
  * (at your option) any later version. 
  * 
  * This program is distributed in the hope that it will be useful, 
  * but WITHOUT ANY WARRANTY; without even the implied warranty of 
  * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the 
  * GNU General Public License for more details. 
  */
#pragma once
```

**Comment by muppetjones** [comment]
@drashna Incorporated your suggestions. Thank you very much for the review!

File: keyboards/lily58/lib/layer_state_reader.c
Comment (muppetjones) [reply keyboards/lily58/lib/layer_state_reader.c]: Restored the original.

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/muppetjones/README.md
Comment (muppetjones) [reply keyboards/lily58/keymaps/muppetjones/README.md:8]:
```suggestion
qmk flash -kb lily58/rev1 -km muppetjones -bl avrdude-split-left
```

File: users/muppetjones/README.md
Comment (muppetjones) [reply users/muppetjones/README.md:1]: Yep!

Comment (muppetjones) [reply users/muppetjones/README.md]: Done.

File: keyboards/lily58/keymaps/muppetjones/config.h
Comment (drashna) [inline keyboards/lily58/keymaps/muppetjones/config.h:26]:
```suggestion
#ifdef OLED_ENABLE
```

**Review by drashna — CHANGES_REQUESTED** [review]

File: keyboards/lily58/keymaps/muppetjones/config.h
Comment (muppetjones) [reply keyboards/lily58/keymaps/muppetjones/config.h]: I feel like this has been changed 20 times -- I'm not sure why it keeps popping back up.

**Comment by muppetjones** [comment]
@drashna g2g

**Review by drashna — APPROVED** [review]

File: keyboards/planck/keymaps/muppetjones/keymap.c
Comment (drashna) [reply keyboards/planck/keymaps/muppetjones/keymap.c:339]: Still needs to be addressed

**Review by drashna — APPROVED** [review]

**Review by drashna — APPROVED** [review]

**Review by keyboard-magpie — APPROVED** [review]

**Comment by keyboard-magpie** [comment]
mis-requested review sorry drashna!

**Comment by muppetjones** [comment]
Thanks!!

**Pull request merged**
